### PR TITLE
Sampler

### PR DESCRIPTION
**Pull request by lonesword**

### CODE REVIEW

File: capreolus/evaluator.py [hunk 1]
```diff
@@ -47,10 +48,7 @@ def eval_runfile(runfile, qrels, metric):
         a dict storing specified metric score and path to the corresponding runfile
     """
     _verify_metric(metric)
-    return {
-        metric: _eval_runfile(runfile, dev_qids=list(qrels.keys()), qrels=qrels, metric=metric),
-        "path": runfile,
-    }
+    return {metric: _eval_runfile(runfile, dev_qids=list(qrels.keys()), qrels=qrels, metric=metric), "path": runfile}
```
Comment (lonesword) [inline capreolus/evaluator.py]: Changes related to black formatting

File: capreolus/extractor/__init__.py [hunk 2]
```diff
@@ -18,3 +18,6 @@ class EmbedText(Extractor):
     @staticmethod
     def config():
         keepstops = False
+
+    def id2vec(self, qid, posdocid, negdocid=None):
```
Comment (lonesword) [inline capreolus/extractor/__init__.py]: Mocking this for now. I'll merge once @Crysitna 's implementation is ready

File: capreolus/sampler/__init__.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
+import random
+import torch.utils.data
+
+from capreolus.utils.loginit import get_logger
+
+
+logger = get_logger(__name__)
+
+
+class TrainDataset(torch.utils.data.IterableDataset):
```
Comment (lonesword) [inline capreolus/sampler/__init__.py]: @andrewyates Do you think this should be a module? I couldn't think of anything that the user might want to configure here, except maybe the seed.

File: capreolus/sampler/__init__.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        qid_to_reldocs = {qid: [docid for docid, label in qrels[qid].items() if label > 0] for qid in qrels}
+        qid_to_negdocs = {qid: [docid for docid, label in qrels[qid].items() if label <= 0] for qid in qrels}
+
+        # Remove the qids that does not have relevant/irrelevant docs associated with it
+        valid_qids = [qid for qid in train_qids if qid_to_reldocs.get(qid) and qid_to_negdocs.get(qid)]
+
+        # Convert each query and doc id to the corresponding feature/embedding and yield
+        def genf():
+            while True:
+                # random.shuffle(valid_qids)
```
Comment (lonesword) [inline capreolus/sampler/__init__.py]: I should set the seed before calling a shuffle right? 
We can have a `self.iteration` variable attached to the dataset class, and then increment it on every yield. Then the seed would be set here as `random.seed(self.iteration)`. The problem here is that `self.iteration` would be the number of times yield has been called - i.e `num_iterations` * `batch_size` - since we increment `self.iterations` on every yield. Not sure if this would be a problem.

File: capreolus/sampler/__init__.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        qid_to_reldocs = {qid: [docid for docid, label in qrels[qid].items() if label > 0] for qid in qrels}
+        qid_to_negdocs = {qid: [docid for docid, label in qrels[qid].items() if label <= 0] for qid in qrels}
+
+        # Remove the qids that does not have relevant/irrelevant docs associated with it
+        valid_qids = [qid for qid in train_qids if qid_to_reldocs.get(qid) and qid_to_negdocs.get(qid)]
+
+        # Convert each query and doc id to the corresponding feature/embedding and yield
+        def genf():
+            while True:
+                random.seed(self.iterations)
```
Comment (lonesword) [inline capreolus/sampler/__init__.py]: `self.iterations` is the number of times a `yield` was made. This number would be `num_iter * batch_size` in fact. This would be a problem if the batch_size changes between 2 experiments/runs. Is that going to be an issue?

File: capreolus/searcher/__init__.py [hunk 6]
```diff
@@ -50,7 +51,8 @@ def _anserini_query_from_file(self, topicsfn, anserini_param_str, output_base_pa
             logger.debug(f"skipping Anserini SearchCollection call because path already exists: {donefn}")
             return
 
-        self["index"].create_index()
+        if not self["index"].exists():
```
Comment (lonesword) [inline capreolus/searcher/__init__.py]: The task is now responsible for making sure that the searcher index is built. Ideally I'd have liked this to happen in the constructor, but as discussed it would end up being annoying in notebooks

File: capreolus/searcher/__init__.py [hunk 7]
```diff
@@ -50,7 +51,8 @@ def _anserini_query_from_file(self, topicsfn, anserini_param_str, output_base_pa
...
+            raise Exception("Index not build")
```
Comment (lonesword) [inline capreolus/searcher/__init__.py]: TODO: Replace this generic exception with a `IndexNotBuild` exception once I pull from @Crysitna 's changes

File: capreolus/task/rerank.py [hunk 8]
```diff
@@ -34,13 +39,36 @@ def describe(config, modules):
...
+
+    topics_fn = benchmark.topic_file
+    metric = "map"
+    searcher.index.create_index()
+    search_results_folder = searcher.query_from_file(topics_fn, os.path.join(searcher.get_cache_path(), benchmark.name))
+    best_search_run_path = evaluator.search_best_run(search_results_folder, benchmark, metric)["path"]
+    best_search_run = searcher.load_trec_run(best_search_run_path)
+    train_dataset = TrainDataset(best_search_run, benchmark, extractor)
+    train_dataloader = torch.utils.data.DataLoader(train_dataset, batch_size=config["batch"])
+    # Call the reranker's score() method here
```
Comment (lonesword) [inline capreolus/task/rerank.py]: We will pass the train_dataloader and the reranker instance to the trainer module here I guess

File: capreolus/sampler/__init__.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        self.extractor = extractor
+
+    def __iter__(self):
+        """
+        Returns: Tuples of the form (query_feature, posdoc_feature)
+        """
+        benchmark = self.benchmark
+        extractor = self.extractor
+        search_run = self.search_run
+        test_qids = benchmark.folds[benchmark.cfg["fold"]]["predict"]["test"]
```
Comment (Crysitna) [inline capreolus/sampler/__init__.py]: I guess we still need a Dataset for dev? or maybe we can put the train/dev/test as an __init__ argument and let them share the genf() code

File: capreolus/sampler/__init__.py (hunk 3 above)
Comment (andrewyates) [reply capreolus/sampler/__init__.py]: no, I think all the (minimal) config options should be passed in. The seed is used across modules, so it makes sense to keep it as a task config option

File: capreolus/sampler/__init__.py (hunk 5 above)
Comment (andrewyates) [reply capreolus/sampler/__init__.py]: That isn't an issue, but I think it doesn't make sense to count inside of the `Dataset`, because this has no advantage over just setting the seed once as we previously. Instead the trainer should set `self.iteration` to reflect the correct iteration, so that the there is no difference between e.g. running the sampler from iteration 1-10 or running it from 1-5 and then later running 5-10. I can modify this to integrate with the trainer though

File: capreolus/sampler/__init__.py (hunk 9 above)
Comment (andrewyates) [reply capreolus/sampler/__init__.py]: I think we should decouple this more from the train/dev/test idea. We can pass in qrels and the valid `(qid, docid)` pairs to be considered, so that the Dataset doesn't have to know anything about the benchmark, searcher, or folds. Then we can have one dataset that samples following some strategy (the training one currently) and one dataset that asks for one prediction for each training instance (the pred one). I'll do this when I integrate with the trainer
